### ISSUE TITLE
Source Object

### ISSUE THREAD

**Issue by jo** [jo/couchdb-bootstrap#43]
I'd like to be able to define a bootstrap layout like this:

```js
module.exports = {
  // set custom config option for CouchDB OS Daemons
  // http://docs.couchdb.org/en/2.1.1/config/externals.html
  _config: {
    migrator: {
      max_jobs: 100
    }
  },
  // setup users Alice and Bob
  _users: {
    alice: {
      _id: 'org.couchdb.user:alice@example.com',
      type: 'user',
      roles: [],
      name: 'alice@example.com',
      password: 'secure'
    },
    bob: {
      _id: 'org.couchdb.user:bob@example.com',
      type: 'user',
      roles: [],
      name: 'bob@example.com',
      password: 'secure'
    }
  },
  // replication jobs from template database to user database
  // http://docs.couchdb.org/en/2.1.1/config/couch-peruser.html
  _replicator: {
    // Alice database
    'todo-app-replication@1:todos-template:userdb-616c696365406578616d706c652e636f6d': {
      source: 'https://couchdb.example.com/todos-template',
      target: 'https://couchdb.example.com/userdb-616c696365406578616d706c652e636f6d',
      continuous: true
    },
    // Bobs database
    'todo-app-replication@1:todos-template:userdb-626f62406578616d706c652e636f6d': {
      source: 'https://couchdb.example.com/todos-template',
      target: 'https://couchdb.example.com/userdb-626f62406578616d706c652e636f6d',
      continuous: true
    }
  },
  // template database, holding indices, views and initial document.
  // continuously replicated to user databases
  'todos-template': {
    // define indices for Mango queries
    // http://docs.couchdb.org/en/2.1.1/api/database/find.html
    _index: {
      'todo-app-index-by-group-id' {
        index: {
          partial_filter_selector: {
            schema: 'todo-item',
            version: 3
           },
          fields: ['schema', 'version', 'groupId']
         },
        ddoc: 'todo-app-index@1',
        name: 'index-by-group-id'
      }
    },
    // install design documents
    _design: {
      'todo-app-view@1': {
        views: {
          todoItems: {
            map: function (doc) {
              if (doc._id.slice(0, 6) === 'todo-item@3:') emit(doc._id.slice(6), doc.isDone)
            },
            reduce: '_count'
          }
        }
      }
    },
    // restrict access to template database
    _security: {
      members: {
        roles: [],
        names: ['admin']
      }
    },
    // initial documents
    'todo-item-group@1:default': {
      title: 'Welcome',
      schema: 'todo-item-group',
      version: 1
    },
    'todo-item@3:welcome': {
      title: 'Getting started with this todo app',
      schema: 'todo-item',
      version: 3,
      groupId: 'default'
    },
    // clean up obsolete versions of design documents and indices
    _delete: [
      'todo-app-index@0.1',
      'todo-app-view@0.1'
    ]
  },
  // setup a database for Alice
  // http://docs.couchdb.org/en/2.1.1/config/couch-peruser.html
  // userdb-{hex encoded alice@example.com}
  'userdb-616c696365406578616d706c652e636f6d': {
    // local docs work well to remember state for services
    // http://docs.couchdb.org/en/2.1.1/api/local.html
    _local: {
      'todo-app-service@1': {
        lastSeq: null,
        status: 'active'
      }
    },
    // only Alice can access db
    _security: {
      members: {
        roles: [],
        names: [
          'alice@example.com'
        ]
      },
      // admin is Alice
      admins: {
        roles: [],
        names: [
          'alice@example.com'
        ]
      }
    }
  },
  // userdb-{hex encoded bob@example.com}
  'userdb-626f62406578616d706c652e636f6d': {
    _local: {
      'todo-app-service@1': {
        lastSeq: null
      }
    },
    // Alice and Bob can access db
    _security: {
      members: {
        roles: [],
        names: [
          'alice@example.com',
          'bob@example.com'
        ]
      },
      // admin is Bob this time
      admins: {
        roles: [],
        names: [
          'bob@example.com'
        ]
      }
    }
  }
}
```

This is currently not possible, bootstrap functions only accept filenames. Since https://github.com/jo/couchdb-compile/releases/tag/v1.10.0 couchdb-compile supports sources being objects.

**Comment by jo** [comment]
**note:** `_index` support will come with #44

**Comment by jo** [comment]
landed in @14.2.0
